### DISCUSSION TITLE
Table not updated

### DISCUSSION THREAD

**Discussion by nghia-vo** [zauberzeug/nicegui#1753] [Q&A] [answered]
### Question

Not sure it's an issue, so I post here. I'm trying to display a 2D array as a table and update the table when the array changed as below:
```python
import numpy as np
from nicegui import ui


def format_table_from_array(data):
    (height, width) = data.shape
    fm_data = np.insert(data, 0, np.arange(height), axis=1)
    columns = [{"name": "column_" + str(j), "label": "Column " + str(j),
                "field": "column_" + str(j), "align": "center"} for j in
               range(width)]
    columns.insert(0, {"name": "id", "label": "Index", "field": "id",
                       "align": "center"})
    rows = []
    for i in range(height):
        dict_tmp = {}
        for j in range(width + 1):
            dict_tmp[columns[j]["name"]] = fm_data[i][j]
        rows.append(dict_tmp)
    return rows, columns


def update_table():
    global rows, columns
    rows, columns = format_table_from_array(np.random.rand(10, 3))
    print("Before: ", table.rows[0])
    table.rows = rows
    print("After: ", table.rows[0])
    table.update()


np.random.seed(0)
rows, columns = format_table_from_array(np.random.rand(10, 3))
ui.button('Update table', on_click=lambda: update_table())
table = ui.table(columns=columns, rows=rows, row_key="id")

ui.run()
```
When click the button, the table doesn't update although the printing message shows the change,
![image](https://github.com/zauberzeug/nicegui/assets/12136997/fc4de463-7fe9-4566-83ad-43b1435cd4f4)

**Comment by falkoschindler** [comment, marked as answer]
The problem occurs because you're replacing the table's `row` attribute with a new object, while the table is still using the old list. It can be easily fixed like this:
```py
table.rows[:] = rows
```

But I just wonder if we could improve the API via properties to avoid this confusion... I'll think about it.

**Reply by falkoschindler** [reply]
See PR #1755.

**Comment by nghia-vo** [comment]
Thanks very much. It works on my example. How about for the case we can't pre-define a table?, i.e. in an app users can choose to display 2d-array as a table or as an image:

```py
rows, columns = None, None
ui.button('Update table', on_click=lambda: update_table())
table = ui.table(columns=columns, rows=rows, row_key="id")
```

**Comment by nghia-vo** [comment]
I've just checked the PR and know that we can do _props['rows'] directly for such case .

```py
if table.rows is None:
    table._props['rows'] = rows
```
